plugins/libplugin-paymake.c

```diff
@@ -0,0 +1,214 @@
+	/* These two paymods must be executed in the exact order
+	 * specified below.
+	 */
+	{"routehints", &routehints_pay_mod},
+	{"presplit", &presplit_pay_mod},
+	{"waitblockheight", &waitblockheight_pay_mod},
+	{"retry", &retry_pay_mod},
+	{"adaptive_splitter", &adaptive_splitter_pay_mod}
+};
+/* Canonical list of paymods involved in MPP.  */
+static const char *const mpp_paymods[] = {
+	"presplit",
+	"adaptive_splitter"
+};
+
+/* Create an initial copy of the paymods.  */
+static struct paymod_desc *
+initial_paymod_list(const tal_t *ctx)
+{
+	struct paymod_desc *rv = tal_dup_arr(ctx, struct paymod_desc,
+					     canonical_paymods,
+					     ARRAY_SIZE(canonical_paymods),
+					     0);
+	for (size_t i = 0; i < tal_count(rv); ++i)
+		rv[i].disabled = false;
+	return rv;
+}
+
+/* From a list of paymods descriptions, create a sequenced list of
+ * paymods.  */
+static struct payment_modifier **
+finalize_paymod_list(const tal_t *ctx,
+		     const struct paymod_desc *paymod_list TAKES)
+{
+	struct payment_modifier **rv = tal_arr(ctx, struct payment_modifier *,
+					       0);
+
+	for (size_t i = 0; i < tal_count(paymod_list); ++i)
+		if (!paymod_list[i].disabled)
+			tal_arr_expand(&rv, paymod_list[i].mod);
+	/* Add NULL terminator.  */
+	tal_arr_expand(&rv, NULL);
+
+	if (taken(paymod_list))
+		tal_free(paymod_list);
+	return rv;
+}
+
+void paymake_global_init(struct plugin *plugin,
+			 const char *buf, const jsmntok_t *t)
+{
+	/* Do nothing.  */
+}
+
+struct paymake {
+	struct command *cmd;
+
+	struct paymod_desc *paymod_list;
+
+	bool amount_fuzz;
+};
+
+struct paymake *paymake_new(const tal_t *ctx, struct command *cmd)
+{
+	struct paymake *pm = tal(ctx, struct paymake);
+	pm->cmd = cmd;
+	pm->paymod_list = initial_paymod_list(pm);
+	pm->amount_fuzz = true;
+	return pm;
+}
+
+struct payment *paymake_create(const tal_t *ctx, struct paymake *pm)
+{
+	struct payment_modifier **mods;
+	struct payment *p;
+	bool remove_fuzz_amount = false;
+
+	if (!pm->paymod_list)
+		/* Misuse of API.  */
+		abort();
+
+	if (!pm->amount_fuzz) {
+		/* Check if the shadowroute paymod is enabled.
+		 * If it is not enabled anyway, no amount fuzzing
+		 * will happen either.  */
+		for (size_t i = 0; i < tal_count(pm->paymod_list); ++i) {
+			if (pm->paymod_list[i].disabled)
+				continue;
+			if (pm->paymod_list[i].mod != &shadowroute_pay_mod)
+				continue;
+			/* Paymod is not disabled and is the shadowroute
+			 * paymod, so we should remove amount fuzzing from
+			 * its data later after construction.  */
+			remove_fuzz_amount = true;
+			break;
+		}
+	}
+
+	mods = finalize_paymod_list(tmpctx, take(pm->paymod_list));
+	pm->paymod_list = NULL;
+
+	p = payment_new(ctx, pm->cmd, NULL, mods);
+	/* Give responsibility of the payment modifiers array to the
+	 * payment.  */
+	tal_steal(p, mods);
+
+	/* Modify the amount_fuzz field of shadowroute, if appropriate.  */
+	if (remove_fuzz_amount)
+		payment_mod_shadowroute_get_data(p)->fuzz_amount = false;
```

Comment:
this seems kind of hacky?

Reply:
That is already how it is currently implemented in current code.  I agree it is hacky, but this seems to be the only way to feed configurations into paymods.